src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,558 @@
+
+                        // Add entry to idToPragmasMap
+                        // Insert the pragmas in reverse order for easier processing later.
+                        if (!idToPragmasMap.TryGetValue(id, out var pragmasForIdInReverseOrder))
+                        {
+                            pragmasForIdInReverseOrder = new List<(SyntaxTrivia pragma, bool isDisable)>();
+                            idToPragmasMap.Add(id, pragmasForIdInReverseOrder);
+                        }
+
+                        pragmasForIdInReverseOrder.Insert(0, (trivia, isDisable));
+                    }
+
+                    if (idsBuilder.Count == 0)
+                    {
+                        // No supported ID in this pragma.
+                        continue;
+                    }
+
+                    hasPragmaInAnalysisSpan = hasPragmaInAnalysisSpan || !span.HasValue || span.Value.OverlapsWith(trivia.Span);
+
+                    sortedPragmasWithIds.Add((trivia, idsBuilder.ToImmutable(), isDisable));
+
+                    // Assume that the prama is not used at the beginning.
+                    pragmasToIsUsedMap.Add(trivia, false);
+                }
+            }
+
+            // Bail out if we have no pragma directives to analyze.
+            if (!hasPragmaInAnalysisSpan)
+            {
+                return;
+            }
+
+            // Compute all the reported compiler and analyzer diagnostics for diagnostic IDs corresponding to pragmas in the tree.
+            var (diagnostics, unhandledIds) = await GetReportedDiagnosticsForIdsAsync(
+                idToPragmasMap.Keys.ToImmutableHashSet(), root, semanticModel, compilationWithAnalyzers,
+                getSupportedDiagnostics, getIsCompilationEndAnalyzer, compilerDiagnosticIds, cancellationToken).ConfigureAwait(false);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Iterate through reported diagnostics which are suppressed in source through pragmas and mark the corresponding pragmas as used.
+            ProcessReportedDiagnostics(diagnostics, tree, compilationWithAnalyzers, idToPragmasMap, pragmasToIsUsedMap);
+
+            // Remove pragma entries for unhandled diagnostic ids.
+            foreach (var id in unhandledIds)
+            {
+                foreach (var (pragma, _) in idToPragmasMap[id])
+                {
+                    pragmasToIsUsedMap.Remove(pragma);
+                }
+            }
+
+            // Finally, report the unnecessary pragmas.
+            ReportUnnecessaryPragmaDiagnostics(pragmasToIsUsedMap, sortedPragmasWithIds,
+                reportDiagnostic, severity, compilationWithAnalyzers.Compilation);
+
+            return;
+
+            static ImmutableArray<string> GetUserExclusions(
+                SyntaxTree tree,
+                CompilationWithAnalyzers compilationWithAnalyzers,
+                CancellationToken cancellationToken,
+                out bool excludeAll)
```

Comment:
consider simplifying this to have the caller get the option and pass it into the simple helper.  i.e. this would be more like `ParseUserExclusions(option,...)`

Reply:
Thanks, that made it much cleaner.